### ISSUE TITLE
New method putDir

### ISSUE THREAD

**Issue by jestersunborn** [theophilusx/ssh2-sftp-client#27]
Would be nice to have a method putDir:

```
sftp.putDir(localDirPath, remoteDirPath)
```

**Comment by jyu213** [comment]
@jestersunborn  
will your want to put all your local directory file to the remote?

**Comment by jestersunborn** [comment]
@jyu213 yes, put the dir with all files inside to the remote dir.

**Comment by ramesh-git** [comment]
Is there any solution provided for this request or is there any workaround to achieve this?

**Comment by luogao** [comment]
want to know is this feature been done, or any other idea? PLZ

**Comment by jestersunborn** [comment]
@luogao I found my very old helper function.

Helper:

```javascript
const recursive = require('recursive-readdir');
const uniq = require('lodash/uniq');

const dirRegExp = /\/(.*\/|)(.*)$/g;

const putDir = (localDir, remoteDir, sftp) =>
  new Promise((resolve, reject) => {
    recursive(localDir, ['.DS_Store'])
      .then(files => (
        files
          .map(file => file.replace(localDir, ''))
          .map(file => file.replace(dirRegExp, `$1`))
          .filter(Boolean)
      ))
      .then(uniq)
      .then(dirs => Promise.all(
        dirs.map(dir => sftp.mkdir(`${remoteDir}/${dir}`, true))
      ))
      .then(() => recursive(localDir, ['.DS_Store']))
      .then(files => (
        files
          .map(file => file.replace(localDir, ''))
          .map(file => file.replace(dirRegExp, `$1`) + file.replace(dirRegExp, `$2`))
      ))
      .then(files => Promise.all(
        files.map(file => sftp.put(`${localDir}/${file}`, `${remoteDir}/${file}`))
      ))
      .then(resolve)
      .catch(err => reject(Error(err)));
  });

module.exports = { putDir  };
```

Usage:
```
putDir(localDirPath, remoteDirPath, sftp);
```

**Comment by luogao** [comment]
@jestersunborn Thanks very muck, it does help a lot 👍 👍 👍

**Comment by jyu213** [comment]
@jestersunborn would you like to pull a request ?

**Comment by Eido95** [comment]
Hope this one will be implemented soon.

**Comment by theophilusx** [comment]
While I do intend to implement this at some point, it is not a high priority. The reality is that implementing such functionality in client code which is specific to that client's needs is fairly trivial, while implementing a generic put/get directory which handles all possible scenarios and platforms is much more complex.  The current focus is to ensure the basic functionality of the module remains correct and has good testing etc. Currently, the focus is to add CI to help ensure the modujle remains reliable and functions correctly. 

This sort of functionality is also on the boarder of what this module does. Essentially, this module provides a high level convenience interface for the ssh2 and ssh2-streams libraries.  Adding put/get directory support goes beyond this and  adds new functionality which is not part of those libraries. Therefore adding this additional functionality may increase maintenance burden and create issues should the underlying libraries change in the future in a way that is incompatible with that functionality. 

Bottom line is that I won't be adding this level of functionality until I'm confident that the testing and CI for the module is robust enough to ensure ongoing stability of the module and  that the additional functionality can be maintained. 

I am certainly open to pullk requests should someone wish to contribute to the implementation of such functionality.

**Comment by theophilusx** [comment]
This functionality has been added for v5.0.0 as uploadDir() and downloadDir(). Hope to release v5.0.0 in next week or so, but in the meantime, people can try out the new functions in the release-5.0.0 branch of the repo.

**Comment by theophilusx** [comment]
This functionality is included in version 5.0.0, which will be released later today
